### PR TITLE
BUG: groupby.agg doesn't include grouping columns in result when selected

### LINKED ISSUE #51186: REF: de-duplicate iterate_slices
```
# From DataFrameGroupBy
    def _iterate_slices(self) -> Iterable[Series]:
        obj = self._selected_obj
        if self.axis == 1:
            obj = obj.T

        if isinstance(obj, Series) and obj.name not in self.exclusions:
            # Occurs when doing DataFrameGroupBy(...)["X"]
            yield obj
        else:
            for label, values in obj.items():
                if label in self.exclusions:
                    continue

                yield values
```

Iterating over self._selected_obj but then excluding self.exclusons seems a lot like just iterating over self._obj_with_exclusions.  I checked that using obj_with_exclusions _does_ break things, but didn't figure out why.

it'd be nice if we could simplify this, or at least document why we need a third thing.  cc @rhshadrach 

If it were just obj_with_exclusions, then in DataFrameGroupBy._indexed_output_to_ndframe we could just use self._obj_with_exclusions.columns instead of re-constructing an Index (i think this re-construction is largely leftover from a time when we dropped nuisance columns)

### Expected Behavior

NA

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #51186 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -1362,7 +1362,7 @@ Groupby/resample/rolling
 - Bug in :meth:`.DataFrameGroupBy.describe` produced incorrect results when data had duplicate columns (:issue:`50806`)
 - Bug in :meth:`.DataFrameGroupBy.agg` with ``engine="numba"`` failing to respect ``as_index=False`` (:issue:`51228`)
 - Bug in :meth:`DataFrameGroupBy.agg`, :meth:`SeriesGroupBy.agg`, and :meth:`Resampler.agg` would ignore arguments when passed a list of functions (:issue:`50863`)
--
+- Bug in :meth:`DataFrameGroupBy.agg` with multiple groupings would not include groupings in the result when they occurred in the selected columns (:issue:`51186`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.0.0.rst:1365]: > with multiple groupings

this might not be clear without looking at the new test

Comment (rhshadrach) [reply doc/source/whatsnew/v2.0.0.rst:1365]: Yea, I wasn't sure how to describe that the bug only occurs with `.groupby([x, y])` and not `.groupby(x)`. Is there better language that could be used here, or maybe just remove it altogether?

Comment (jbrockmendel) [reply doc/source/whatsnew/v2.0.0.rst:1365]: 'after subsetting columns (e.g. `gb[["a", "b"]].agg(...)`'?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1320,21 +1320,15 @@ def aggregate(self, func=None, *args, engine=None, engine_kwargs=None, **kwargs)
...
+        if isinstance(obj, Series):
             # Occurs when doing DataFrameGroupBy(...)["X"]
             yield obj
         else:
             for label, values in obj.items():
-                if label in self.exclusions:
-                    # Note: if we tried to just iterate over _obj_with_exclusions,
-                    #  we would break test_wrap_agg_out by yielding a column
-                    #  that is skipped here but not dropped from obj_with_exclusions
-                    continue
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1336]: so can we replace usages (there's just one) with obj_with_exclusions (maybe transpose)?

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1336]: OK for follow-up

**Comment by jbrockmendel** [comment]
Potential follow-up that this makes possible: _iterate_slices is only used in _python_agg_general, where it iterates over columns calling agg_series.  agg_series iterates over groups.  Instead we could iterate differently over groups, which would look a lot like _aggregate_frame.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by rhshadrach** [comment]
Assuming no objection, going to merge when greenish. Currently targeted at 2.1.

**Pull request merged**